### PR TITLE
fix `build.rs`: Use env `RUSTC` if present

### PR DESCRIPTION
**Pull request by NobodyXu**
- Use `env::var_os` to obtain `OUT_DIR` since it could contains non utf-8 characters.
 - Improve `write_host_rs` readability using raw string.
 - Flush the `out: File` in `write_host_rs` before returning.

Signed-off-by: Jiahao XU <Jiahao_XU@outlook.com>

### CODE REVIEW

File: build.rs [hunk 1]
```diff
@@ -87,137 +90,90 @@ fn write_host_rs(mut out: File, triple: Triple) -> io::Result<()> {
...
+        operating_system = triple.operating_system,
+        environment = triple.environment,
+        binary_format = triple.binary_format,
     )?;
-    writeln!(out, "        }}")?;
-    writeln!(out, "    }}")?;
-    writeln!(out, "}}")?;
 
-    Ok(())
+    out.flush()
```
Comment (sunfishcode) [inline build.rs:176]: It isn't necessary to `flush()` a `File` because `File`s are unbuffered and their `flush()` doesn't do anything.

**Review by sunfishcode — COMMENTED** [review]
I'm curious; have you ever seen a build environment with a non-Unicode directory name, or is `var_os` here just theoretical correctness?

File: build.rs (hunk 1 above)
Comment (NobodyXu) [reply build.rs]: Thanks, I've reverted this change.

File: build.rs [hunk 2]
```diff
@@ -58,15 +60,16 @@ fn main() {
...
-                } else {
-                    return None;
-                }
-            }
-            _ => return None,
+        let rustc = if let Some(rustc) = env::var_os("RUSTC") {
+            Cow::Owned(PathBuf::from(rustc))
+        } else {
+            Cow::Borrowed(Path::new("rustc"))
+        };
```
[the remark this answers is not in the document]
Comment (NobodyXu) [reply build.rs]: Thanks, I've fixed it according to your suggestion.

**Comment by NobodyXu** [comment]
> I'm curious; have you ever seen a build environment with a non-Unicode directory name, or is `var_os` here just theoretical correctness?

Well, no, I haven't seen one, but given that it is possible to do that on Linux and the filesystem can accept any zero-terminated `[u8]`, I think it's only time until some users try to build with non-utf8 characters in their directory name.

Since fixing this only involves one-line change to `env::var_os`, I think it won't hurt to fix this issue.

**Comment by sunfishcode** [comment]
Indeed, it doesn't look like it will hurt in this instance. But similar things do hurt in several projects I'm involved in, so I'm always curious who actually needs non-Unicode paths and why, and whether all the complexity that we pass on to everyone for their benefit is worthwhile.

**Comment by NobodyXu** [comment]
> But similar things do hurt in several projects I'm involved in, so I'm always curious who actually needs non-Unicode paths and why, and whether all the complexity that we pass on to everyone for their benefit is worthwhile.

Can you provide me with a few example of these cases please?
I guess it's because having to use `PathBuf` instead of `String` adds complexity and forces people to learn yet another string-type, plus it's not compatible with `String`/`str` and all the other custom string types like `CompactString`.

**Pull request merged**